src/main/java/org/wildfly/security/credential/store/CredentialStoreSpi.java

```diff
@@ -125,4 +143,79 @@ public boolean isInitialized() {
+                return cs.retrieve(commandSpec, credentialType);
+            } catch (NoSuchAlgorithmException | UnsupportedCredentialTypeException e) {
+                throw new CredentialStoreException(e);
+            }
+        } else if (trimmedCommandSpec.startsWith("{" + CmdPasswordStore.SUPPORTED_CMD_TYPE)) {
+            try {
+                CredentialStore cs = CredentialStore.getInstance(CmdPasswordStore.CMD_PASSWORD_STORE);
+                return cs.retrieve(commandSpec, credentialType);
+            } catch (NoSuchAlgorithmException | UnsupportedCredentialTypeException e) {
+                throw new CredentialStoreException(e);
+            }
+        } else if (trimmedCommandSpec.startsWith(MaskedPasswordStore.PASS_MASK_PREFIX)) {
+            try {
+                CredentialStore cs = CredentialStore.getInstance(MaskedPasswordStore.MASKED_PASSWORD_STORE);
+                cs.initialize(attributes);
+                return cs.retrieve(commandSpec, credentialType);
+            } catch (NoSuchAlgorithmException | UnsupportedCredentialTypeException e) {
+                throw new CredentialStoreException(e);
+            }
+        } else if (trimmedCommandSpec.startsWith("{" + CUSTOM_PROVIDER)) {
+            try {
+                StringTokenizer tokenizer = new StringTokenizer(commandSpec, "{}");
+                String providerSpec = tokenizer.nextToken();
+                StringTokenizer providerTokenizer = new StringTokenizer(providerSpec, "@");
+                String providerType = providerTokenizer.nextToken();
+                CredentialStore execCS = CredentialStore.getInstance(providerType);
+                execCS.initialize(attributes);
+                return execCS.retrieve(commandSpec, credentialType);
+            } catch (NoSuchAlgorithmException | UnsupportedCredentialTypeException e) {
+                throw new CredentialStoreException(e);
+            }
+        } else {
+            try {
+                return credentialType.cast(createPasswordCredential(commandSpec.toCharArray()));
+            } catch (NoSuchAlgorithmException | InvalidKeySpecException e) {
+                throw new CredentialStoreException(e);
+            }
+        }
+    }
+
+    /**
+     * Utility method which creates {@link PasswordCredential} based on {@link ClearPasswordSpec} using {@code password} parameter.
+     * @param password to create {@link PasswordCredential} from
+     * @return {@link PasswordCredential} created using {@link ClearPasswordSpec}
+     * @throws NoSuchAlgorithmException when {@link PasswordFactory} doesn't know password algorithm (should not happen as we are using our own stuff)
+     * @throws InvalidKeySpecException when {@link PasswordFactory#generatePassword(KeySpec)} is using wrong key spec. (should not happen as we are using our own stuff)
+     */
+    protected PasswordCredential createPasswordCredential(char[] password) throws NoSuchAlgorithmException, InvalidKeySpecException {
```

Comment:
This method definitely doesn't belong here, since you can just do this instead:

``new PasswordCredential(ClearPassword.createRaw("clear", password));``.

This is a general purpose SPI so having a special credential-handling shortcut method doesn't really fit in.